### ISSUE TITLE
check that fedmsg isn't already initialized before initializing

### ISSUE THREAD

**Issue by ralphbean** [fedora-infra/trac-fedmsg-plugin#1]
The symptom we're seeing is that after a while of running, trac starts
reporting "unable to find an available endpoint"; this is coming from the
fedmsg code.

So, with this endpoints issue (presuming we're just using mod_wsgi which I
think is the case now) one of the processes is complaining that it can't bind
to any of the ports we told it to.

mod_wsgi is configured there to use 4 threads and 4 processes, so.. 16 total
threads.  and.. I gave it 16 endpoints to bind to.

So, it seems like trac internally is doing something I didn't expect it to,
like, reinitializing some of its plugins after a while.  this would cause the
plugin to try and re-bind to a new port, but since it has the old one already
claimed.. it "runs out", hence "couldn't find an available endpoint".  I could
add some code to the fedmsg-trac plugin making it check that it doesn't already
have a socket floating out there somewhere before it tries to set itself up
again.  :(

If anyone can do that before I get to it, I will totally give you a badge.

**Comment by ralphbean** [comment]
Hm, it doesn't seem possible for fedmsg to be initializing more than once per thread.  There's a check in place already (see https://github.com/fedora-infra/trac-fedmsg-plugin/blob/develop/trac_fedmsg_plugin.py#L72).

I'll try doubling the number of endpoints and see if that makes a difference.

**Comment by ralphbean** [comment]
Well, nevermind.  We're in alpha freeze right now and I'm not sure if this will actually work, so let's hold off until we're not in freeze.

**Comment by ralphbean** [comment]
Okay, we're in between the alpha and beta freezes right now.  I've increased the number of endpoints with the following puppet patch:

``` diff
commit 9bb40081e6f59bbe387b25a69393030e1b988706
Author: Ralph Bean <rbean@redhat.com>
Date:   Fri Oct 11 12:51:24 2013 +0000

    Declare 32 fedmsg endpoints for fedorahosted.

diff --git a/modules/fedmsg/templates/fedmsg.d/endpoints.py.erb b/modules/fedmsg/templates/fedmsg.d/endpoints.py.erb
index 7dd7f17..67370b6 100644
--- a/modules/fedmsg/templates/fedmsg.d/endpoints.py.erb
+++ b/modules/fedmsg/templates/fedmsg.d/endpoints.py.erb
@@ -111,13 +111,18 @@ config = dict(

 <% if environment != 'staging' %>
         # fedorahosted trac runs as 4 processes with 4 threads each.
+        # that would normally mean 16 endpoints, but we were getting some
+        # unexpected errors where users would intermittently be shown tracebacks
+        # indicating that there were not enough endpoints.  As an attempted fix,
+        # we'll declare 32 endpoints here.  If we get a report about that
+        # traceback again, then we should look more closely.
         "trac.hosted03": [
             "tcp://hosted03.%s:30%02i" % (vpn_suffix, i)
-            for i in range(16)
+            for i in range(32)
         ],
         "trac.hosted04": [
             "tcp://hosted04.%s:30%02i" % (vpn_suffix, i)
-            for i in range(16)
+            for i in range(32)
         ],
 <% end %>
```

**Comment by ralphbean** [comment]
Ok.  I haven't heard any complaints about this in a month, so I'm closing the ticket.

**Comment by ralphbean** [comment]
Got another report https://fedorahosted.org/rel-eng/ticket/5818

**Comment by ralphbean** [comment]
This is going to require digging into the trac source code to see what its doing with its plugin management.

**Comment by ralphbean** [comment]
Just a note that this is blocking https://fedorahosted.org/fedora-badges/ticket/98

**Comment by ralphbean** [comment]
This should be fixed in `trac-fedmsg-plugin-0.3.0`.
